### PR TITLE
`WaitForSnapshotStep` verifies if the index belongs to the latest snapshot of that SLM policy

### LINKED ISSUE #57809: "Wait for snapshot" action in Delete phase of ILM doesn't check if index was backed up
Hello team,

while adding a field for "[wait for snapshot](https://www.elastic.co/guide/en/elasticsearch/reference/master/ilm-wait-for-snapshot.html) policy" to Delete phase in Index Lifecycle Management UI, I noticed that this action does not in fact ensure that a snapshot of the index exists before deleting the index. This can lead to irreversible data loss of documents in the managed index.

How to recreate this behaviour:

1. Create a repository for snapshots
2. Create a snapshot policy that backs up an index
3. Create a **different** index (this is the **managed** index) with an alias for rollover
4. Create an index lifecycle policy that deletes the managed index with "wait for snapshot" option
5. After conditions for Delete phase are met and snapshot policy is executed, the managed index is deleted. This index can't be restored as the snapshot contains a different index. 

<details>
<summary>Expand for console commands</summary>

```
PUT /_snapshot/my_repo
{
  "type": "fs",
  "settings": {
    "location": "./my_repo_test",
    "compress": true
  }
}
```

(index to be backed up `my_snapshot_index`, snapshots created every minute and deleted after 10 min)
```
PUT _slm/policy/my_snapshot_policy
{
  "name": "<snapshot-{now}>",
  "schedule": "0 * * * * ?",
  "repository": "my_repo",
  "config": {
    "indices": [
      "my_snapshot_index"
    ]
  },
  "retention": {
    "expire_after": "10m",
    "min_count": 1,
    "max_count": 3
  }
}

PUT /my_test_index-1
PUT /my_test_index-1/_alias/my_alias
```

(rollover after 3 docs, delete after `my_snapshot_policy` created a snapshot).
```
PUT _ilm/policy/my_policy
{
  "policy": {
    "phases": {
      "hot": {
        "min_age": "0ms",
        "actions": {
          "rollover": {
            "max_size": "50gb",
            "max_age": "30d",
            "max_docs": 3
          },
          "set_priority": {
            "priority": 100
          }
        }
      },
      "delete": {
        "min_age": "0d",
        "actions": {
          "wait_for_snapshot": {
            "policy": "my_snapshot_policy"
          },
          "delete": {}
        }
      }
    }
  }
}
```
</details>

### PR DESCRIPTION
**Pull request by gmarouli**
The `WaitForSnapshotStep` used to check if the SLM policy has been executed after the index has entered the delete phase, but it did not check if the SLM policy included this index.

The result of this is that if the user used an SLM policy that did not include this index, when the index would enter the `WaitForSnapshotStep`, it would wait for a snapshot to be taken, a snapshot that would not include the index, and then ILM would delete the index.

See the exact reproduction path: https://github.com/elastic/elasticsearch/issues/57809

**Solution**
This PR, after it finds a successful SLM run, it verifies if the snapshot taken by SLM contains this index. If not it throws an error, otherwise it proceeds.

ILM explain will report:
```
"step_info": {
        "type": "illegal_state_exception",
        "reason": "the last successful snapshot of policy 'hourly-snapshots' does not include index '.ds-my-other-stream-2023.10.16-000001'"
      }
```

**Backwards compatibility concerns**
In this PR, the `WaitForSnapshotStep` changed from `ClusterStateWaitStep` to `AsyncWaitStep`. We do not think this is gonna cause an issue. This was tested manually by the following steps:
- Run a master node with the old version.
- When ILM is executing `wait-for-snapshot`, we shutdown the node
- We start the node again with the new version os ES
- ES was able to pick up the step and continue with the new code.

We believe that this covers bwc concerns.

Fixes: https://github.com/elastic/elasticsearch/issues/57809

### CODE REVIEW

**Review by joegallo — COMMENTED** [review]
I think the code mostly looks very good. I've added a couple of admittedly trivial comments.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForSnapshotStep.java [hunk 1]
```diff
@@ -79,15 +90,37 @@ public Result isConditionMet(Index index, ClusterState clusterState) {
...
+        String snapshotName = snapPolicyMeta.getLastSuccess().getSnapshotName();
+        String repositoryName = snapPolicyMeta.getPolicy().getRepository();
+        GetSnapshotsRequest request = new GetSnapshotsRequest().snapshots(new String[] { snapshotName })
+            .repositories(repositoryName)
+            .includeIndexNames(true)
+            .verbose(false);
+        getClient().admin().cluster().getSnapshots(request, ActionListener.wrap(response -> {
+            if (response.getSnapshots().size() != 1) {
+                listener.onFailure(
+                    new IllegalStateException(
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForSnapshotStep.java:111]: Sharp, yes, I missed this one.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForSnapshotStep.java [hunk 2]
```diff
@@ -79,15 +90,37 @@ public Result isConditionMet(Index index, ClusterState clusterState) {
...
         logger.debug(
             "executing policy because snapshot start time {} is after action time {}, snapshot timestamp is {}",
             snapPolicyMeta.getLastSuccess().getSnapshotStartTimestamp(),
             actionTime,
             snapPolicyMeta.getLastSuccess().getSnapshotFinishTimestamp()
         );
-        return new Result(true, null);
+        String snapshotName = snapPolicyMeta.getLastSuccess().getSnapshotName();
+        String repositoryName = snapPolicyMeta.getPolicy().getRepository();
+        GetSnapshotsRequest request = new GetSnapshotsRequest().snapshots(new String[] { snapshotName })
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForSnapshotStep.java:104]: Good point, this is the order in the endpoint too. I will fix it.

**Comment by gmarouli** [comment]
Thank you for review @joegallo, you keep me sharp 🤓 .

**Comment by gmarouli** [comment]
About the backport labels, I think we should try to backport it, I do realise that backporting to `7.17.x` might be tricky, that's why I would like to timebox it at least give it a try see if it's worth it.

**Comment by joegallo** [comment]
Regarding backporting, I think it should be pretty straightforward (hopefully!), a lot of the ILM code has been pretty stable for a while. A tricky bit is that there's no generally available `EmptyInfo` on 7.17, I think that was introduced pretty recently via #100179. Maybe it's worth pre-gaming a `>non-issue` PR that only introduces the `EmptyInfo` like #100179 did?

**Review by joegallo — APPROVED** [review]
🚀

**Comment by gmarouli** [comment]
@elasticmachine update branch

**Pull request merged**